### DISCUSSION TITLE
Disable MultiTenant process in certain circumstances

### DISCUSSION THREAD

**Discussion by Glursh** [Finbuckle/Finbuckle.MultiTenant#798] [Q&A]
Hi everybody

In our ASP.NET Core application (.NET 7, Finbuckle.MultiTenant 6.12.0), in certain circumstances, we have to deal with a bulk data processing (CRUD) that involve multiple tenants.

Can we "just turn off" MultiTenant during this particular process?

**Comment by AndrewTriesToCode** [comment]
Hi, in your db context do you have explicit tenant id columns or are you utilizing the shadow properties that Finbuckle will add in MultiTenantDbContext?

**Reply by Glursh** [reply]
Actually, we don't have an explicit TenantId, so we use the shadow properties, yes.
We are now thinking to make a fork of your great projects with this kind of code :

in interface IMultiTenantDbContext :

```
public interface IMultiTenantDbContext
{
    ITenantInfo TenantInfo { get; }
    TenantMismatchMode TenantMismatchMode { get; }
    TenantNotSetMode TenantNotSetMode { get; }
    bool IsMultiTenantEnable { get; } // <--------------
}
```

in abstract class MultiTenantDbContext :

```
public bool IsMultiTenantEnable { get; set; } = true; // <--------------

...

public override int SaveChanges(bool acceptAllChangesOnSuccess)
{
    if (IsMultiTenantEnable) // <--------------
    {
        this.EnforceMultiTenant();
    }
    return base.SaveChanges(acceptAllChangesOnSuccess);
}

public override async Task<int> SaveChangesAsync(bool acceptAllChangesOnSuccess, CancellationToken cancellationToken = default(CancellationToken))
{
    if (IsMultiTenantEnable) // <--------------
    {
        this.EnforceMultiTenant();
    }
    return await base.SaveChangesAsync(acceptAllChangesOnSuccess, cancellationToken);
}
```

in FinbuckleEntityTypeBuilderExtensions

```
// build expression tree for EF.Property<string>(e, "TenantId") == TenantInfo.Id'
var predicate = Expression.Equal(leftExp, rightExp);

var forkContextIsMultiTenantEnableExp = Expression.Property(contextMemberAccessExp, nameof(IMultiTenantDbContext.IsMultiTenantEnable)); // <--------------
// build expression tree for "IsMultiTenantEnable == False"
var forkPredicate = Expression.Equal(forkContextIsMultiTenantEnableExp, Expression.Constant(false)); // <--------------
// build expression tree for "IsMultiTenantEnable == False || EF.Property<string>(e, "TenantId") == TenantInfo.Id"
predicate = Expression.OrElse(forkPredicate, predicate); // <--------------

// combine with existing filter
if (existingQueryFilter != null)
{
    predicate = Expression.AndAlso(existingQueryFilter.Body, predicate);
}

// build the final expression tree
var delegateType = Expression.GetDelegateType(builder.Metadata.ClrType, typeof(bool));
var lambdaExp = Expression.Lambda(delegateType, predicate, entityParamExp);


```

**Reply by AndrewTriesToCode** [reply]
Yeah I see what you mean.

On your context set TenantMismatchMode to ignore and it’ll let you save and update anything you want. If you need to query you can use the IgnoreQueryFilters LINQ method to ignore tenant filter on results. That will probably get you what you need for this operation. 

Ive also had some success looping through each tenant, creating an instance of the db context for that tenant and making changes, then continue the loop. Works ok but can be slow. I will have factory methods to better support creating for a tenant soon.  I can share some code if you’d like. 

Another alternative is doing it via straight sql if that’s an option.

**Reply by Glursh** [reply]
Thanks for your help.

Using IgnoreQueryFilters it's a good idea, but it may break other features, like soft delete.

We had some issues with TenantMismatchMode, as it doesn't handle some specific cases.
For example, to update certain properties without first retrieving the model in database :
```
var e = new MyModel() { Id = ***, ModifiedOn = ***, WhateverProperty = *** }; // where MyModel has the MultiTenantAttribute
dbContext.MyModels.Attach(e);
var entry = dbContext.Entry(e);
entry.State = EntityState.Modified;
entry.Property("TenantId").IsModified = false;
await dbContext.SaveChangesAsync();
```

With this code, the EnforceMultiTenant() method will try to set TenantId to the current TenantInfo.Id value, even with TenantMismatchMode.Ignore :
```
// handle Tenant Id not set for modified entities
var notSetModified = modifiedMultiTenantEntities.
    Where(e => (string?)e.Property("TenantId").CurrentValue == null).ToList();
if (notSetModified.Any())
{
    switch (tenantNotSetMode)
    {
        case TenantNotSetMode.Throw:
            throw new MultiTenantException($"{notSetModified.Count} modified entities with Tenant Id not set.");

        case TenantNotSetMode.Overwrite:
            foreach (var e in notSetModified)
            {
                e.Property("TenantId").CurrentValue = tenantInfo.Id;
            }
            break;
    }
}
```

**Reply by AndrewTriesToCode** [reply]
There’s a subtle difference there between tenant mismatch and tenant not set (which I guess is a form of mismatch technically). I can’t recall offhand but can you set a shadow property value to what you want it to be so that it is set by the time that check runs?

Also I do recommend adding explicit tenant id fields if you can, much easier to deal with these situations where you can just set it to what you want on the model. I know it’s not always possible though.

**Reply by Glursh** [reply]
Precisely, without going into details, it happens that we only want to modify, in a bulk process, certain properties in different Tenants, without having to worry about the Tenant of each record.

Anyway, thanks for your advices and answers. 😊
We'll create and use a fork of your project to achieve our goals.
